### ISSUE TITLE
Advisory `--advisory-severities` and type filters produce union instead of intersection

### ISSUE THREAD

**Issue by pkratoch** [rpm-software-management/dnf5#1401]
For example, `dnf5 advisory list --security --advisory-severities=critical` shows all advisories that are either or type security or have critical severity (so even e.g. moderate securities or critical bugfixes are shown).

I believe that other filter combinations produce intersection (except when specifying multiple type filters, but that makes sense if viewed as if it was just one list option), so this is inconsistent. If there is a reason for this behavior, it should be at least documented.

**Comment by ppisar** [comment]
I confirm the bug. --advisory-severities option is document as "limit…", but it does not do an intersection:

~~~~
# dnf5 advisory list --security --advisory-severities=critical
Updating and loading repositories:
Repositories loaded.
Name                   Type     Severity                           Package              Issued
FEDORA-2024-43a0920f12 security Low      perl-Clipboard-0.29-1.fc39.noarch 2024-04-10 14:44:52
~~~~

**Comment by kontura** [comment]
I think the reason is compatibility with old dnf which is unfortunately also doing an union.

I agree that the docs should definitely be fixed (both man pages and `--help`).
I have also noticed that the `--help` of advisory commands refers to packages which is confusing because its filtering advisories. This is probably caused by the fact that the options are shared with other commands.

**Comment by j-mracek** [comment]
I have created a patch, but I am not sure whether we should modify the behavior or modify DNF5 in more complex way.

I think that both following command should target the same set of advisories.

`dnf5 advisory list --security --advisory-severities=critical`

`dnf5 upgrade --security --advisory-severities=critical`

What do you think?

**Comment by ppisar** [comment]
Both commands should behave the same.

I'm for changing it to an intersection because users who want an union can use the options in separate calls. Contrary with the union behavior, users cannot not achieve the intersection.

**Comment by kontura** [comment]
I agree that all the commands should behave the same.

I also prefer intersection but changing it now could be quite unexpected for users.

**Comment by j-mracek** [comment]
> I agree that all the commands should behave the same.
> 
> I also prefer intersection but changing it now could be quite unexpected for users.

@kontura May I ask you when you would prefer that the change happen?

**Comment by kontura** [comment]
> @kontura May I ask you when you would prefer that the change happen?

I think the best time was when we introduced the advisory options to dnf5 but since the change was rejected at that time due to compatibility concerns I wanted to mention it.

**Comment by j-mracek** [comment]
I am still not sure whether the change make sense from `upgrade` command usages.

I think that command `dnf5 upgrade --security --advisory-severities=critical` with a result that both filters are applied with `and` operator os OK and expected.

But other combination might be difficult `dnf5 upgrade --advisory FEDORA-2024-1706127797 --security`
- It returns updates for advisory `FEDORA-2024-a7a8f8f01b` and all other security updates
- The new behavior results in an empty set because the advisory has the `Bugfix` type.
- It means that after the change **several combination of filters would be allowed but not usable**.

Other combination might be difficult `dnf5 upgrade --advisory FEDORA-2024-1706127797 --bzs 2271998`
- It returns updates for advisory `FEDORA-2024-a7a8f8f01b` and bugzilla 2271998
- The new behavior results in an empty set because the advisory does not fix the Bugzilla.

The same problem is related to CVEs.

What about to only modify applying and operator between type and severity?
Or we can add a restrictions that only some combination will be not allowed.

**Comment by ppisar** [comment]
I will repeat myself, I like the new behavior more.

> What about to only modify applying and operator between type and severity?

Does severity exists in non-security advisories?

**Comment by j-mracek** [comment]
Yes, severity is used with other types of advisories, but not so often
```
===============================================================================
  Update ID: FEDORA-2024-39f1aba503
       Type: bugfix
    Updated: 2024-05-01 03:36:32
       Bugs: 2276226 - python-identify-2.5.36 is available
Description: 2.5.36
   Severity: Low
```

**Comment by m-blaha** [comment]
Although the DNF team agrees that using intersection instead of the currently implemented union would have been a better choice, we’ve decided not to change the command behavior at this stage. Maintaining compatibility (even with dnf4 in this case) is currently our priority. We will improve the documentation, as it currently suggests that intersection is used, which can be misleading.
